File: MatchingGame/MatchingGame.m
//
//  MatchingGame.m
//  MatchingGame
//
//  Created by CSSE Department on 4/30/20.
//  Copyright © 2020 CSSE Department. All rights reserved.
//

#import "MatchingGame.h"

@implementation MatchingGame

- (id) init{
    return [self initWithNumPairs: 10];
}

- (id) initWithNumPairs:(NSInteger) numPairs {
    self = [super init];
    if (self) {
        self.numPairs = numPairs;
        NSArray* allCardBacks = [@"🎆,🎇,🌈,🌅,🌇,🌉,🌃,🌄,⛺,⛲,🚢,🌌,🌋,🗽" componentsSeparatedByString:@","];
        NSArray* allEmojiCharacters = [@"🚁,🐴,🐇,🐢,🐱,🐌,🐒,🐞,🐫,🐠,🐬,🐩,🐶,🐰,🐼,⛄,🌸,⛅,🐸,🐳,❄,❤,🐝,🌺,🌼,🌽,🍌,🍎,🍡,🏡,🌻,🍉,🍒,🍦,👠,🐧,👛,🐛,🐘,🐨,😃,🐻,🐹,🐲,🐊,🐙" componentsSeparatedByString:@","];


        // Randomly select emojiSymbols
        NSMutableArray* emojiSymbolsUsed = [[NSMutableArray alloc] init];
        while (emojiSymbolsUsed.count < numPairs) {
            NSString* symbol = allEmojiCharacters[arc4random_uniform((UInt32) allEmojiCharacters.count)];
            if (![emojiSymbolsUsed containsObject:symbol]) {
                [emojiSymbolsUsed addObject:symbol];
            }
        }
        [emojiSymbolsUsed addObjectsFromArray:emojiSymbolsUsed];
        // Shuffle the NSMutableArray before converting it to an NSArray.
        for (int i = 0; i < emojiSymbolsUsed.count; ++i) {
            UInt32 j = arc4random_uniform((UInt32) emojiSymbolsUsed.count - i) + i;
            [emojiSymbolsUsed exchangeObjectAtIndex:i withObjectAtIndex:j];
        }
        self.cards = [NSArray arrayWithArray:emojiSymbolsUsed];

        // Randomly select a card back.
        self.cardBack = allCardBacks[arc4random_uniform((UInt32) allCardBacks.count)];

        // Reset cardStates to ensure default values.
        for (int i = 0; i < self.cards.count; ++i) {
            cardStates[i] = CardStateHidden;
        }
    }
    self.firstCard = -1;
    self.secondCard = -1;
    return self;
}

- (bool) pressedCardAtIndex:(NSInteger)index{
    if(self.gameState == GameStateGameWon || self.gameState == GameStateTurnComplete){
        return false;
    }
    if(cardStates[index] != CardStateHidden){
        return false;
    }
    if(self.gameState == GameStateFirstTurn){
        self.firstCard = index;
        cardStates[index] = CardStateShowing;
        self.gameState = GameStateSecondTurn;
        return true;
    }
    if(self.gameState == GameStateSecondTurn){
        self.secondCard = index;
        cardStates[index] = CardStateShowing;
        self.gameState = GameStateTurnComplete;
        return true;
    }
    return false;
}

- (bool) startNewTurn{
    if(self.firstCard == -1 || self.secondCard == -1){
        NSLog(@"Incomplete turn");
        return false;
    }
    if(self.cards[self.firstCard] == self.cards[self.secondCard]){
        cardStates[self.firstCard] = CardStateRemoved;
        cardStates[self.secondCard] = CardStateRemoved;
        self.gameState = GameStateFirstTurn;
        self.firstCard = -1;
        self.secondCard = -1;
        return [self checkWin];
    }else{
        cardStates[self.firstCard] = CardStateHidden;
        cardStates[self.secondCard] = CardStateHidden;
        self.gameState = GameStateFirstTurn;
        self.firstCard = -1;
        self.secondCard = -1;
        return false;
    }
}

- (bool) checkWin{
    for(int i = 0; i < self.cards.count; ++i){
        if(cardStates[i] != CardStateRemoved) return false;
    }
    self.gameState = GameStateGameWon;
    return true;
}

- (NSString*) description{
    NSString* str = @"";
    for(int i = 0; i < self.cards.count; ++i){
        str = [str stringByAppendingFormat:@"%@", self.cards[i]];
        if((i + 1) % 4 == 0){
            str = [str stringByAppendingString: @"\n"];
        }
    }
    return str;
}

- (cardState) getCardStateAtIndex:(NSInteger) index{
    return cardStates[index];
}

@end

File: MatchingGame/MatchingGame.swift
//
//  MatchingGame.swift
//  MatchingGame
//
//  Created by CSSE Department on 4/14/20.
//  Copyright © 2020 CSSE Department. All rights reserved.
//

import Foundation
class MatchingGame : CustomStringConvertible{
    
    let allCardBacks = Array("🎆🎇🌈🌅🌇🌉🌃🌄⛺⛲🚢🌌🌋🗽")
    let allEmojiCharacters = Array("🚁🐴🐇🐢🐱🐌🐒🐞🐫🐠🐬🐩🐶🐰🐼⛄🌸⛅🐸🐳❄❤🐝🌺🌼🌽🍌🍎🍡🏡🌻🍉🍒🍦👠🐧👛🐛🐘🐨😃🐻🐹🐲🐊🐙")
    
    enum cardState : String{
        case hidden = "Hidden", showing = "Showing", removed = "Removed"
    }
    
    enum gameStates : String{
        case firstTurn, secondTurn, turnComplete, gameWon
        
        func simpleDescription() -> String{
            switch self{
            case .firstTurn:
                return "Waiting for first selection"
            case .secondTurn:
                return "waiting for second selection"
            case .turnComplete:
                return "Turn complete"
            case .gameWon:
                return "The game is over"
            }
        }
    }
    
    var cards : [Character]
    var cardBack : Character
    var cardStates : [cardState]
    var gameState : gameStates
    var firstCard : Int?
    var secondCard : Int?
    
    init(){
    // Randomly select emojiSymbols
        var emojiSymbolsUsed = [Character]()
        while emojiSymbolsUsed.count < 10 {
          let index = Int(arc4random_uniform(UInt32(allEmojiCharacters.count)))
          let symbol = allEmojiCharacters[index]
          if !emojiSymbolsUsed.contains(symbol) {
            emojiSymbolsUsed.append(symbol)
          }
        }
        cards = emojiSymbolsUsed + emojiSymbolsUsed
        cards.shuffle()

        // Randomly select a card back for this round
        let index = Int(arc4random_uniform(UInt32(allCardBacks.count)))
        cardBack = allCardBacks[index]
        gameState = .firstTurn
        cardStates = [cardState](repeating: .hidden, count: 20)
        firstCard = nil
        secondCard = nil
    }
    
    func pressedCard(atIndex : Int) -> Bool{
        if gameState == gameStates.gameWon || gameState == gameStates.turnComplete {
            return false
        }
        if cardStates[atIndex] == .removed || cardStates[atIndex] == .showing {
            return false
        }
        if gameState == .firstTurn {
            firstCard = atIndex
            cardStates[atIndex] = .showing
            gameState = .secondTurn
            return true
        }
        if gameState == .secondTurn {
            secondCard = atIndex
            cardStates[atIndex] = .showing
            gameState = .turnComplete
            return true
        }
        return false
    }
    
    func startNewTurn() -> Bool{
        if firstCard == nil || secondCard == nil {
            print("Incomplete turn")
            return false
        }
        if cards[firstCard!] == cards[secondCard!] {
            cardStates[firstCard!] = .removed
            cardStates[secondCard!] = .removed
            gameState = .firstTurn
            firstCard = nil
            secondCard = nil
            return checkWin()
        }else{
            cardStates[firstCard!] = .hidden
            cardStates[secondCard!] = .hidden
            gameState = .firstTurn
            firstCard = nil
            secondCard = nil
            return false
        }
    }
    
    func checkWin() -> Bool {
        for state in cardStates{
            if state != .removed {
                return false
            }
        }
        gameState = .gameWon
        return true
    }
    
    var description: String {
        get{
            var str = ""
            for index in 0..<cards.count{
                str += String(cards[index])
                if (index + 1) % 4 == 0 {
                    str += "\n"
                }
            }
            return str
        }
    }
    
    

}

// Helper method to randomize the order of an Array. See usage later.  Copied from:
// http://stackoverflow.com/questions/24026510/how-do-i-shuffle-an-array-in-swift
extension Array {
    mutating func shuffle() {
        for i in 0..<(count - 1) {
            let j = Int(arc4random_uniform(UInt32(count - i))) + i
            self.swapAt(i, j)
        }
    }
}
